[MLIR] Implement remark emitting policies in MLIR

grypp opened the pull request:
This update introduces two new remark emitting policies:
1. `RemarkEmittingPolicyAll`, which emits all remarks,
2. `RemarkEmittingPolicyFinal`, which only emits final remarks after processing.

The `RemarkEngine` is modified to support these policies, allowing for more flexible remark handling based on user configuration.

PR also adds flag to `mlir-opt`
```
  --remark-policy=<value>                                    - Specify the policy for remark output.
    =all                                                     -   Print all remarks
    =final                                                   -   Print final remarks
```

Relanding https://github.com/llvm/llvm-project/pull/160526

This PR requires RemarkEngine to be finalize manually. So here is usage:
```
MLIRContext ctx;
ctx.setRemarkEngine(...)
...
ctx.getRemarkEngine().shutdown() <-- PR adds this, it is required when the emission policy is final
```

grypp commented on mlir/include/mlir/IR/Remarks.h:
```diff
@@ -406,9 +429,15 @@ class RemarkEngine {
 
+  /// Shutdown the remark engine. This will finalize the remark engine and
+  /// close the output file.
+  void shutdown();
```
When the emissions policy is final, we now defer emitting remarks until the very end.
In my previous PR, postponed remarks were emitted during MLIRContext destruction. That turned out to be problematic, since it required accessing the MLIRContext while it was being destroyed.

In this PR, I introduced a finalize() function that must be called explicitly before destroying the MLIRContext.

I’m happy to adjust if there’s a better approach.

joker-eph replied:
I would really want to avoid this kind of explicit step exposed to the users, this is an indication of a problem in the design in general.

Can the MLIRContext destructor handle the finalization?

grypp replied:
> Can the MLIRContext destructor handle the finalization?

I tried, but it doesn’t work. If there were a destructor I could call there, that would be great.  

The problem is that during `~MLIRContext`, `RemarkEngine` needs to access `MLIRContext` to get location infor and use `mlir::emitRemark`. At that point, `DiagnosticEngine` has already erased its handler, so no handler is left.

grypp committed:
[MLIR] Implement remark emitting policies in MLIR

This update introduces two new remark emitting policies:
1. `RemarkEmittingPolicyAll`, which emits all remarks,
2. `RemarkEmittingPolicyFinal`, which only emits final remarks after processing.

The `RemarkEngine` is modified to support these policies, allowing for more flexible remark handling based on user configuration.

PR also adds flag to `mlir-opt`
```
  --remark-policy=<value>                                    - Specify the policy for remark output.
    =all                                                     -   Print all remarks
    =final                                                   -   Print final remarks
```

grypp committed:
fx

razvanlupusoru reviewed (approved):
Thank you Guray! Looks reasonable to me :)
